test/test_internet.py

```diff
@@ -9,22 +9,61 @@
 import pytest
 import warnings
 from urllib.error import URLError
+from urllib.request import Request
 
 import torchvision.datasets.utils as utils
 from common_utils import get_tmp_dir
 
 
-class TestDatasetUtils:
+@pytest.fixture
+def patch_url_redirection(mocker):
+    class Response:
+        def __init__(self, url):
+            self.url = url
+
+    def factory(*urls):
+        class PatchedOpener:
+            def __init__(self, request_or_url, *args, **kwargs):
+                self._request_or_url = request_or_url
+
+            def __enter__(self):
+                url = (
+                    self._request_or_url.full_url
+                    if isinstance(self._request_or_url.full_url, Request)
+                    else self._request_or_url.full_url
+                )
```

Comment:
Looks like this is `self._request_or_url.full_url` in all cases?
Also, maybe we can slightly simplify some logic consididering we know we'll always pass an URL, not a Request object?

Reply:
Yeah sorry. I have a habit of making stuff more generic and in turn complicated in places where it is not needed.